flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -538,36 +538,45 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
 									coordinatorsToCheckpoint, pendingCheckpoint, timer),
 							timer);
 
-			CompletableFuture.allOf(masterStatesComplete, coordinatorCheckpointsComplete)
-				.whenCompleteAsync(
-					(ignored, throwable) -> {
-						final PendingCheckpoint checkpoint =
-							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
-
-						if (throwable == null && checkpoint != null && !checkpoint.isDiscarded()) {
-							// no exception, no discarding, everything is OK
-							final long checkpointId = checkpoint.getCheckpointId();
-							snapshotTaskState(
-								timestamp,
-								checkpointId,
-								checkpoint.getCheckpointStorageLocation(),
-								request.props,
-								executions,
-								request.advanceToEndOfTime);
-
-							coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
-
-							onTriggerSuccess();
-						} else {
-								// the initialization might not be finished yet
-								if (checkpoint == null) {
-									onTriggerFailure(request, throwable);
+			FutureUtils.assertNoException(
```

Comment:
Won't this kill JVM on _any_ exception regardless of whether it was handled below or not?

Reply:
You are right, we have to change `whenCompleteAsync` into `handleAsync`. I will update the PR.

Reply:
Thanks for the update.

I'm also not sure about `assertNoException` which calls `System.exit` internally. It can cause problems because:
- there might be other jobs
- depending on the setup, it could be problematic to find the reason; e.g. buffered logs can be lost
- we skip any cleanup

Instead, we could notify `CheckpointFailureManager` so that it would terminate only this job.
What do you think?

Reply:
I think that any error which comes out of the `handleAsync` call is a programming error which is not recoverable. Hence, I believe that we should fail hard as we also do it if there is an uncaught exception in a thread, for example.

Reply:
It's a bit different for a thread because there we know absolutely nothing about the failure and can't fail a relevant job. But this is already a philosophical discussion I think :)

Reply:
I agree that `FutureUtils.assertNoException` is a bit rough here, however it would be more complex to fail the job through `CheckpointFailureManager`. Because we have to consider two more things,
1. Does the checkpoint is properly initialized? `CheckpointFailureManager` requires a checkpoint id but we might not have one.
2. `isTriggering` must be taken care of. Failing job does not reset this flag.

So `FutureUtils.assertNoException` is acceptable for me. Technically speaking, it should never be reached unless there is a bug of our code or some other extreme corner cases, like OOM.

Reply:
`FutureUtils.assertNoException` is similar to `assert` or `Preconditions.checkState`, they all ensure that certain invariants are kept. In our case it is that no exception is being thrown by the `handleAsync` callback.